packages/plugin-git/sources/gitUtils.ts

```diff
@@ -317,3 +318,98 @@ export async function clone(url: string, configuration: Configuration) {
     return directory;
   });
 }
+
+export async function fetchRoot(initialCwd: PortablePath) {
+  // Note: We can't just use `git rev-parse --show-toplevel`, because on Windows
+  // it may return long paths even when the cwd uses short paths, and we have no
+  // way to detect it from Node (not even realpath).
+
+  let match: PortablePath | null = null;
+
+  let cwd: PortablePath;
+  let nextCwd = initialCwd;
+  do {
+    cwd = nextCwd;
+    if (await xfs.existsPromise(ppath.join(cwd, `.git` as Filename)))
+      match = cwd;
+    nextCwd = ppath.dirname(cwd);
+  } while (match === null && nextCwd !== cwd);
+
+  return match;
+}
+
+export async function fetchBase(root: PortablePath, {baseRefs}: {baseRefs: Array<string>}) {
+  if (baseRefs.length === 0)
+    throw new UsageError(`Can't run this command with zero base refs specified.`);
+
+  const ancestorBases = [];
+
+  for (const candidate of baseRefs) {
+    const {code} = await execUtils.execvp(`git`, [`merge-base`, candidate, `HEAD`], {cwd: root});
+    if (code === 0) {
+      ancestorBases.push(candidate);
+    }
+  }
+
+  if (ancestorBases.length === 0)
+    throw new UsageError(`No ancestor could be found between any of HEAD and ${baseRefs.join(`, `)}`);
+
+  const {stdout: mergeBaseStdout} = await execUtils.execvp(`git`, [`merge-base`, `HEAD`, ...ancestorBases], {cwd: root, strict: true});
+  const hash = mergeBaseStdout.trim();
+
+  const {stdout: showStdout} = await execUtils.execvp(`git`, [`show`, `--quiet`, `--pretty=format:%s`, hash], {cwd: root, strict: true});
+  const title = showStdout.trim();
+
+  return {hash, title};
+}
+
+// Note: This returns all changed files from the git diff, which can include
+// files not belonging to a workspace
+export async function fetchChangedFiles(root: PortablePath, {base, project}: {base: string, project: Project}) {
+  const {stdout: localStdout} = await execUtils.execvp(`git`, [`diff`, `--name-only`, `${base}`], {cwd: root, strict: true});
+  const trackedFiles = localStdout.split(/\r\n|\r|\n/).filter(file => file.length > 0).map(file => ppath.resolve(root, npath.toPortablePath(file)));
+
+  const {stdout: untrackedStdout} = await execUtils.execvp(`git`, [`ls-files`, `--others`, `--exclude-standard`], {cwd: root, strict: true});
+  const untrackedFiles = untrackedStdout.split(/\r\n|\r|\n/).filter(file => file.length > 0).map(file => ppath.resolve(root, npath.toPortablePath(file)));
+
+  return [...new Set([...trackedFiles, ...untrackedFiles].sort())];
+}
```

Comment:
Shouldn't `fetchChangedFiles` include the `changesetIgnore` logic from the original function?:

https://github.com/yarnpkg/berry/pull/3459/files#diff-218fec89381039f75ec127dc4b415f20ff81968c878433ea1920ff774f179414R33-R41

Reply:
Ah yep. I left it out originally as I wasn't sure how much code should be shared with the config options that were introduced for the version plugin.

Fixed in 711507b73bf4475858bdcd05d4d4aa4bb6bd5335.